smartapps/curb/curb-energy-manager.src/curb-energy-manager.groovy

```diff
@@ -0,0 +1,248 @@
+        count += count
+        } catch (e) {
+          log.debug(e);
+        }
+
+    }
+
+    //log.debug(devlist)
+
+    def sorted = devlist.sort { a, b -> b.pct <=> a.pct }
+    def message = "Curb Alert: Energy usage is projected to go over selected threshold."
+    //log.debug(devlist.size())
+    if (devlist.size() > 3)
+    {
+      message = "Curb Alert: Energy usage is projected to go over selected threshold. Your biggest consumers currently are: ${sorted[0].name} ${sorted[0].pct}%, ${sorted[1].name} ${sorted[1].pct}%, and ${sorted[2].name} ${sorted[2].pct}%"
+    }
+  sendNotificationToContacts(message, recipients)
+}
+
+def resetClocking() {
+  //log.debug("resetting the clock")
+  state.readings = []
+  for (int i = 0; i < Integer.parseInt(timeInterval); i++) {
+    state.readings[i] = null
+  }
+  state.usage = 0
+  //log.debug(state)
+  if (state.throttling == true){
+    stopThrottlingUsage()
+  }
+}
+
+def runAutomation() {
+  if (enabled){
+    def mf = new java.text.SimpleDateFormat("m")
+    def minute = Integer.parseInt(mf.format(new Date())) % Integer.parseInt(timeInterval)
+
+      //log.debug("in automation / minute: " + minute.toString())
+
+      // First Minute, reset the whole process
+    if (minute == 0) {
+      resetClocking()
+    }
+
+      //we're still in the projection period. don't do anything
+      if (minute < Float.parseFloat(timeInterval) * (Float.parseFloat(projectionPeriod) / 100) )
+      {
+        return
+      }
+
+    state.usage = 0.0
+    def samples = 0.0
+    for (int i = 0; i < Integer.parseInt(timeInterval); i++) {
+      if (state.readings[i] != null) {
+        samples = samples + 1.0
+        state.usage = state.usage + (state.readings[i] / 60 / 1000)
+      }
+    }
+    //log.debug("samples:" + samples.toString())
+    //log.debug("usage:" + state.usage.toString())
+
+    if (samples != 0.0) {
+        def avgedUsage = minute * ( state.usage / samples )
+          //log.debug(avgedUsage)
+          def safetyThreshold = ( Float.parseFloat(kwhThreshold) * ( 1 - (Float.parseFloat(safetyMargin) / 100)))
+          //log.debug(safetyThreshold)
+        if (avgedUsage > safetyThreshold) {
+        throttleUsage()
+      }
+    }
+  }
+}
+
+def checkEnergyMonitor(evt) {
+  if (!checkRunning()) {
+    return
+  }
+
+  def mf = new java.text.SimpleDateFormat("m")
+  def minute = Integer.parseInt(mf.format(new Date())) % Integer.parseInt(timeInterval)
+
+  def power = meter.currentState("power").value
+  state.readings[minute] = Float.parseFloat(power)
+  //log.debug(state)
+}
+
+def captureContollerStates() {
+  if (!state.throttling) {
+    for (t in thermostats) {
+      state[t.id] = t.currentState("thermostatMode").value
+    }
+    for (s in switches) {
+      state[s.id] = s.currentState("switch").value
+    }
+    //log.debug "state: ${state}"
+  }
+}
+
+def restoreControllerStates() {
+  if (!state.throttling) {
```

Comment:
This method is only called from stopThrottlingUsage() after `state.throttling` has been set to false, so is this check necessary?